### DISCUSSION TITLE
To find angle of pixels in 1 bit tiff file

### DISCUSSION THREAD

**Discussion by vkkindia** [libvips/libvips#2837] [Help wanted]
I have a properity software which tells the angle of the pixels wrt top left as well as lpi in the 1 bit tiff file.

is it possible to use a libvips function for this. I tried reading the functions but could not get a function related to this.

Also how to count the vertical black pixels in the file. i tried vips countlines, but the output is not correct. my idea is to get the "lpi" (yes lines per inch not dpi) for that tiff file to cross verify the created output.

Thanks for your wondurful software.

**Comment by jcupitt** [comment]
Hi, could you explain what you mean in more detail? Do you want to find the image skew? A sample image plus sample output would help.

**Reply by vkkindia** [reply]
Hi sir,

I have attached two images one tif file that needs to be checked for dot angle and lpi.
another image is the output of proprietary software for your reference.  
For our convenience i have zoomed in, the left top dot box left most side to see how the dots are formed. 

thanks sir.
[https://drive.google.com/file/d/1V3jAEyAfoG0Kdh-4YbzPWb559IgLj4F4/view?usp=sharing](url)

**Reply by jcupitt** [reply]
Ah, you want to estimate the angle of the halftone screen?

Just take the FFT, find the peak nearest the origin, and calculate the angle and pythagorean distance.

**Reply by vkkindia** [reply]
sir, how to do it in python, similarly how to get that lpi,

**Reply by jcupitt** [reply]
It's `image.fwfft()` for the forward FFT, then use `max` to find the position of the peak, then a little trig to get the angle and distance.

**Comment by vkkindia** [comment]
Hi sir,

I compiled with the ffw library to enable this option.
This is the first time i am working in fft
I read few documentation on fft, had an opportunity to explore that.
###############
import cv2 as cv
import numpy as np
from matplotlib import pyplot as plt
img = cv.imread('C_cropped_01x01.tif',0)
f = np.fft.fft2(img)
fshift = np.fft.fftshift(f)
magnitude_spectrum = 20*np.log(np.abs(fshift))
plt.subplot(121),plt.imshow(img, cmap = 'gray')
plt.title('Input Image'), plt.xticks([]), plt.yticks([])
plt.subplot(122),plt.imshow(magnitude_spectrum, cmap = 'gray')
plt.title('Magnitude Spectrum'), plt.xticks([]), plt.yticks([])
plt.show()
################


*using libvips*

###########################
#/usr/bin/python3

import pyvips

input_filename="C_cropped_01x01.tif"
ftt_out_filename="fourier.jpg"
im = pyvips.Image.new_from_file(input_filename)

im_fft = im.fwfft()
im_fft.write_to_file(ftt_out_filename)

im_fft_max = im_fft.max()
print("im_max is ", im_fft_max)

###################

Actually thereafter I dont know how to proceed, totally new in FFT, how to
calculate for angle and lpi.

thanks again

**Comment by jcupitt** [comment]
Hi, I get "page not found" for your file link.

**Comment by vkkindia** [comment]
sir
,
Pls try now
https://drive.google.com/file/d/1V3jAEyAfoG0Kdh-4YbzPWb559IgLj4F4/view?usp=sharing

**Reply by jcupitt** [reply]
I had a quick go in nip2:

![image](https://user-images.githubusercontent.com/580843/172370652-4e37f424-8826-43f9-8424-316913db86d7.png)

I don't know how accurate the result is. You'd need to experiment with different size tiles, and maybe doing a slight blur before looking for the peak. Maybe sample from a range of random positions and take the average. 

Workspace here: http://www.rollthepotato.net/~john/detect-screen.ws

**Comment by vkkindia** [comment]
Thanks sir, i will check that.
I also found a page related to my question,

https://stackoverflow.com/questions/24259251/detect-angle-of-a-black-white-halftone-screen

**Comment by vkkindia** [comment]
Sir,

I have tested this script in nip2, but not getting desired results.
uploaded few test files
https://drive.google.com/file/d/198WGU24764omOqbZpqfDbbxd3lOg7Oxv/view?usp=sharing
the filename mentions the angle and lpi detected by other software. Can you
check that sir.

thanks

**Comment by vkkindia** [comment]
Sir,

Pls find attached the example1.gs which will create different halftone and
angles, the proprietary software find angles and lpi as mentioned in ps.
https://drive.google.com/file/d/1ILB-rGu3enWTldlAGXimr-5dnr3C8X_I/view?usp=sharing

pls use latest ghostscript
https://ghostscript.com/releases/gsdnld.html

ver gs-9-56 is very fast, earlier version buggy create black dots in empty
white which will affect the calculation.

~/Downloads/gs/ghostscript-9.56.1-linux-x86_64/gs-9561-linux-x86_64 -dBATCH
-dSAFER -dNOPAUSE -dNumRenderingThreads=4 -r1440 -sOutputFile=test%d.tif
-sDEVICE=tiffsep1 -f example1.ps

pls test the created **(Black).tif*
